autoPyTorch/data/tabular_feature_validator.py

```diff
@@ -517,11 +525,17 @@ def infer_objects(self, X: pd.DataFrame) -> pd.DataFrame:
                     self.logger.warning(f'Casting the column {key} to {dtype} caused the exception {e}')
                     pass
         else:
-            # Calling for the first time to infer the categories
-            X = X.infer_objects()
-            for column, data_type in zip(X.columns, X.dtypes):
-                if not is_numeric_dtype(data_type):
-                    X[column] = X[column].astype('category')
+            if len(self.dtypes) != 0:
+                # when train data has no object dtype, but test does
+                # we prioritise the datatype given in training data
+                for column, data_type in zip(X.columns, self.dtypes):
+                    X[column] = X[column].astype(data_type)
+            else:
+                # Calling for the first time to infer the categories
+                X = X.infer_objects()
+                for column, data_type in zip(X.columns, X.dtypes):
+                    if not is_numeric_dtype(data_type):
+                        X[column] = X[column].astype('category')
```

Comment:
```suggestion
        elif len(self.dtypes) != 0:  # when train data has no object dtype, but test does
            # we prioritise the datatype given in training data
            for column, data_type in zip(X.columns, self.dtypes):
                X[column] = X[column].astype(data_type)
        else:  # Calling for the first time to infer the categories
            X = X.infer_objects()
            for column, data_type in zip(X.columns, X.dtypes):
                if not is_numeric_dtype(data_type):
                    X[column] = X[column].astype('category')
```

Reply:
I think these are just preferences on where to start the comment.

Reply:
Nono, it actually removed an indent level.

Reply:
actually, if you notice we are also saving the dtypes in  `self.object_dtype_mapping` which should be done for both of the two conditions you moved back an indent level. So, I think its fine the way it is.

Reply:
Oh yeah, I did not notice, but I also did not notice that we still have the same issue (which happens when we have a huge number of features) in this method.
Could you fix it?

```python
        if hasattr(self, 'object_dtype_mapping'):
            # Mypy does not process the has attr. This dict is defined below
            try:
                X = X.astype(self.object_dtype_mapping)
            except Exception as e:
                self.logger.warning(f'Casting the column {key} to {dtype} caused the exception {e}')
                pass
            return

        if len(self.dtypes) != 0:
            # when train data has no object dtype, but test does.  Prioritise the datatype given in training data
            dtype_dict = {col: dtype for col, dtype in zip(X.columns, self.dtypes)}
            X = X.astype(self.object_dtype_mapping)
        else:
            # Calling for the first time to infer the categories
            X = X.infer_objects()
            dtype_dict = {col: 'category' for col, dtype in X.items() if not is_numeric_dtype(dtype)}
            X = X.astype(dtype_dict)

        # only numerical attributes and categories
        self.object_dtype_mapping = {col: dtype for col, dtype in zip(X.columns, X.dtypes)}
```